### ISSUE TITLE
Where is Network/Page commands defined in this library ?

### ISSUE THREAD

**Issue by GuacheSuede** [cyrus-and/chrome-remote-interface#233]
eg.
   ` Network.requestWillBeSent((params) => {
        console.log(params.request.url);
    });`
    `Page.loadEventFired(() => {
        client.close();
    });`

Where are these 2 functions defined in the JS files ? Could not find anything from searching.
Thanks.

**Comment by GuacheSuede** [comment]
Or rather a better question after inspection would be, how do you map it from the api ?

**Comment by cyrus-and** [comment]
The JavaScript API is [built][1] from the local [`protocol.json`][2] file, from the one fetched from Chrome or even from a user-provided file.

In this case `Network.requestWillBeSent` and `Page.loadEventFired` are just events, i.e., not commands. Commands are translated into JSON messages and sent to Chrome, while events and command responses are sent back from Chrome, converted to proper objects, *demultiplexed* and returned back to the user via the API.

In other words the implementation is not here, this module is just a Node.js interface to the WebSocket protocol used by Chrome.

I hope that this answers your questions.

[1]: https://github.com/cyrus-and/chrome-remote-interface/blob/fc870f24fe148b76f7be5a4a9dd9f6551005fc52/lib/api.js#L59-L81
[2]: https://github.com/cyrus-and/chrome-remote-interface/blob/master/lib/protocol.json

**Comment by GuacheSuede** [comment]
Having a better understanding now. Would i be right to say `Network.requestWillBeSent` is sent via [here](https://github.com/cyrus-and/chrome-remote-interface/blob/master/lib/chrome.js#L75) as params ?

**Comment by GuacheSuede** [comment]
Apologies, am trying to visualize the process of how the `Network.requestWillBeSent` function is sent via the websocket to chrome.

**Comment by cyrus-and** [comment]
> Would i be right to say Network.requestWillBeSent is sent via here as params ?

Nope, the `send` method is used to send commands while `Network.requestWillBeSent` is an event; and events are processed [here][1].

The process is more or less like this:
1. events are enabled for a certain domain, e.g., `Network.enable()` (a message is sent through the WebSocket using the `send` method that you linked);
2. the client registers using `Network.requestWillBeSent`;
3. when the event triggers, Chrome sends a message through the WebSocket and it's handled by the aforementioned [piece of code][1] which is responsible to emit the Node.js event that is ultimately used by the client.

> Perhaps could these just be shorthand for this ?

Not exactly.

```js
Network.requestWillBeSent((params) => { doSomethingWith(params); })
```

is equivalent to:

```js
chrome.on('Network.requestWillBeSent', (params) => { doSomethingWith(params); });
```

While

```js
chrome.once('Network.requestWillBeSent', (params) => { doSomethingWith(params); });
```

is equivalent to:

```js
const params = await Network.requestWillBeSent;
doSomethingWith(params);
```

[1]: https://github.com/cyrus-and/chrome-remote-interface/blob/fc870f24fe148b76f7be5a4a9dd9f6551005fc52/lib/chrome.js#L300-L304

**Comment by GuacheSuede** [comment]
@cyrus-and Apologies for late reply was out of town. Alright sure, Thanks for the help alot
